utils.groovy

```diff
@@ -482,6 +496,65 @@ def run_cloud_tests(pipecfg, stream, version, s3_stream_dir, basearch, commit) {
     parallel testruns
 }
 
+// Runs followup upgrade tests based on conditions
+def run_upgrade_tests(pipecfg, stream, version, basearch, commit) {
+    def stream_info = pipecfg.streams[stream]
+
+    def min_supported_start_versions = [
+        aarch64: 34,
+        ppc64le: 40, // We haven't released on ppc64le yet
+        s390x:   36,
+        x86_64:  32,
+    ]
+
+    // Because of an x86_64 bootloader issue the oldest that is supported
+    // on x86_64 is 32.x: https://github.com/coreos/fedora-coreos-tracker/issues/1448
+    // All other arches were shipped later than 32 so a min_version of 32 works.
+    def min_version = 32 as Integer
+    // Use a maximum version that matches the current stream Fedora major version
+    def manifest = readYaml(text:shwrapCapture("cosa shell -- cat src/config/manifest.yaml"))
+    def max_version = manifest.releasever as Integer
+
+    // A list of all the start versions we want to run tests for
+    def start_versions = []
+
+    // If this is a production build we'll run a test from the
+    // earliest release of each Fedora major we've supported for this
+    // stream. Iterate over the possibilities and kick off a job for
+    // each.
+    if (stream_info.type == 'production') {
+        for (start_version in min_version..max_version) {
+            start_versions += start_version
+        }
+    }
+
+    // If this is a non-production build we'll run a single test from
+    // a start version determined by a modulo of the day of the month.
+    // This prevents running the full matrix for each development
+    // build, but gradually covers the full matrix over a period of time.
+    if (stream_info.type in ['development', 'mechanical']) {
+        def now = new Date()
+        def dayOfMonth = now.getAt(Calendar.DAY_OF_MONTH)
+        def offset = dayOfMonth % (min_version - max_version)
```

Comment:
Using days for this means that the same starting point will be used for multiple builds in the same day. Also if the pipeline is broken for a day, we won't get coverage for that starting point until the next cycle.

WDYT instead of something like:


```suggestion
        def builds = readJSON(file: "builds/builds.json")
        def offset = builds.size % (max_version - min_version + 1)
```

?